File: Editorial/107_Binary_Tree_Level_Order_Traversal_II/Go/1_Recursion_DFS_Preorder_Traversal.go
func levelOrderBottom(root *TreeNode) [][]int {
    levels := make([][]int, 0)
    var helper func(node *TreeNode, level int)
    helper = func(node *TreeNode, level int) {
        if node == nil {
            return
        }
        // start the current level
        if len(levels) == level {
            levels = append(levels, []int{})
        }
        // append the current node value
        levels[level] = append(levels[level], node.Val)
        // process child nodes for the next level
        if node.Left != nil {
            helper(node.Left, level+1)
        }
        if node.Right != nil {
            helper(node.Right, level+1)
        }
    }
    helper(root, 0)
    // reverse the levels
    for i := 0; i < len(levels)/2; i++ {
        levels[i], levels[len(levels)-1-i] = levels[len(levels)-1-i], levels[i]
    }
    return levels
}

File: Editorial/107_Binary_Tree_Level_Order_Traversal_II/TypeScript/1_Recursion_DFS_Preorder_Traversal.ts
function levelOrderBottom(root: TreeNode | null): number[][] {
    let levels: number[][] = [];
    function helper(node: TreeNode | null, level: number) {
        if (!node) return;
        // start the current level
        if (!levels[level]) levels[level] = [];
        // append the current node value
        levels[level].push(node.val);
        // process child nodes for the next level
        if (node.left) helper(node.left, level + 1);
        if (node.right) helper(node.right, level + 1);
    }
    helper(root, 0);
    return levels.reverse();
}
